src/model.lisp

```diff
@@ -52,7 +52,7 @@
 
 
 
-(defclass model (rl-model 3d-object)
+(defclass model (3d-object)
```

Comment:
I don't agree that this is the right fix. All throughout the code various RL-* classes are the parents of their Claylib-native (non-RL) variants and the design assumes that, even if only implicitly.

Instead, look to the texture classes. In texture.lisp we have: `rl-texture`, `texture`, `texture-object`, and a `tex` mixin (may or may not be needed here).

Now, models are higher up the food chain than textures, but think about how we're using assets and objects. In an MMO you'd still probably have many player characters sharing the same model with different textures attached. So IMO a `rl-model`/`model`/`model-object` scheme still makes sense. Admittedly I'm a little rusty on the class hierarchy, but does that sound reasonable?

Reply:
Hmm, I'll need to think more about this. I get what you're saying about the texture stuff - I basically made `model` the equivalent of `texture-object`. But I'm not sure about some things. For example sharing models between characters and attaching different textures: to set the texture for a model you would be modifying the underlying C struct (with `SetMaterialTexture`). I'm probably missing something but wouldn't we need new `rl-model`s for each character? Come to think of it, how would Raylib do this? I'll investigate.

Reply:
To be fair, this will almost certainly present its own set of challenges, but it's still important to get right. I can dive in to help if you need it. In the end we may have to build our own loading and drawing functions by calling Raylib's mesh and material functions rather than directly using `DrawModelEx` and the like.

Reply:
I was _almost_ ready to say that we could probably ignore Raylib's Model class and build an equivalent ourselves out of meshes, materials, and animation data. But in practice that's starting to look like it could be a pain, too.

First of all, there's no `LoadMesh(es)` function that I can find, so you have to `LoadModel` regardless. That gets you all the data at least, but as you pointed out, it's not necessarily easy to use efficiently in that format -- there's not even a `CopyModel` function! So then we pull that data into our own object(s) and can use `DrawMesh` and maybe `DrawMeshInstanced` for most of our drawing. But... what about animations? Well, AFAICT, you can't animate a "list of Meshes," you have to animate a Model. Okay, so we... recreate a Model from our Meshes via `LoadModelFromMesh`?? Except -- you guessed it -- that function only takes _one_ mesh, so we'd have to merge them except by doing so we'd probably mess up the materials (and no, Raylib doesn't have a `MergeMeshes` either), and at this point it's too late because I've already thrown my computer out the window.

So, for now, we keep using Model, I think. Looking forward to another perspective on what happens beyond that.

Reply:
Actually, the animation problem should be really easy to solve. Take a look: https://github.com/raysan5/raylib/blob/master/src/rmodels.c#L1854

`UpdateModelAnimation` only uses the model to get the list of meshes and a bind pose; aside from that it just animates the meshes as you'd expect. So we could write our own version of `UpdateModelAnimation` that takes that information directly. Could be in Lisp but it's just as well to patch Raylib and push it upstream if they'll accept.

If we do that, the issue about having to load a model and then pick it apart becomes relatively minor, I think. Then for our own version of instancing, we have `model-object`s which use `mesh-asset`s, `material-asset`s, and `animation-asset`s with their own `transform`s.

Thoughts?

Reply:
> `UpdateModelAnimation` only uses the model to get the list of meshes and a bind pose

Similar kind of thing also applies to `DrawModelEx`: https://github.com/raysan5/raylib/blob/master/src/rmodels.c#L3274

Reply:
Okay I slept on this, came back to it, and now it makes sense. Restating:

**Extracting data**: We use `LoadModel` & `LoadModelAnimations` to get our Lisp objects (`mesh-asset`, `material-asset`, `animation-asset`).

**Instancing**: Create a `model-object` which reuses the above data, but initializes its own `transform` (I think backed by a Raylib `Matrix` rather than pure Lisp). Though we still have to copy the underlying material data when we make a `model-object`, otherwise `SetMaterialTexture` will overwrite it. We can do that with a `copy-asset-to-object` method which wraps a contribution to Raylib: `CopyMaterial` or something. Luckily we don't need to copy animations because each `model-object` can just track the particular frame and animation it's on.

**Drawing/Animating**: From here we specialize `draw-object` on `model-objects` which could wrap another contribution, `DrawMeshes`, just `DrawModelsEx` which takes the meshes directly. Same idea for `UpdateAnimation` & `UpdateModelAnimation`, but our lisp interface would probably look something like this:
```lisp
(defun update-animation (model-object n frame)
  (claylib/ll:update-animation (c-asset (meshes model-object)) 
                               (c-asset (nth n (animations model-object))) 
                               frame))
```

Reply:
I think you have the right idea. In general we need to watch out for anything that tries to update a Model's fields, because Raylib doesn't offer any (easy) way to distinguish between a loaded model asset and a drawn model object. I was hoping we could just write our own version of `DrawModelEx` and not need to worry about `UpdateAnimation` but I don't think that's going to be possible.